model_templates/gpu_nemo_textgen/custom_model/custom.py

```diff
@@ -1,25 +1,106 @@
+import logging
 import os
+import subprocess
+from pathlib import Path
+from subprocess import CalledProcessError
+
 import boto3
 from mlpiper.extra.aws_helper import AwsHelper
-from datarobot_drum import RuntimeParameters
 
+from datarobot_drum import RuntimeParameters
+from datarobot_drum.custom_task_interfaces.user_secrets import SecretType
+from datarobot_drum.drum.enum import LOGGER_NAME_PREFIX
+from datarobot_drum.drum.exceptions import DrumCommonException
+from datarobot_drum.drum.gpu_predictors import get_optional_parameter
 
 # this path is required by NeMo Inference Server
 NEMO_MODEL_STORE_DIR = "/model-store"
 
 
 def load_model(code_dir: str):
-    print("Downloading model from S3...")
+    load_model_response = (
+        "succeeded"  # a non-empty response is required to signal that load_model succeeded
+    )
+
+    s3_url = get_optional_parameter("s3Url")
+    if s3_url:
+        s3_credential = RuntimeParameters.get("s3Credential")
+        s3_client = S3Client(s3_url, s3_credential)
+
+        keys, dirs = s3_client.list_objects()
+        s3_client.download_files(NEMO_MODEL_STORE_DIR, keys, dirs)
+        return load_model_response
+
+    ngc_registry_url = get_optional_parameter("ngcRegistryUrl")
+    if ngc_registry_url:
+        ngc_credential = RuntimeParameters.get("ngcCredential")
+        ngc_client = NGCRegistryClient(ngc_credential)
+        ngc_client.download_model_version(ngc_registry_url)
+        return load_model_response
+
+    raise Exception(
+        "Can't download model artifacts. The 'ngcRegistryUrl' or 's3Url'"
+        " are expected to be set in the Runtime Parameters."
+    )
+
+
+class NGCRegistryClient:
+    logger = logging.getLogger(LOGGER_NAME_PREFIX + "." + __name__)
+
+    def __init__(self, ngc_credential):
+        if ngc_credential["credentialType"] != SecretType.API_TOKEN.value:
```

Comment:
I think you can specify the credential type in the definition now and let the platform do the validation for you.

Reply:
:thinking: 
I see this trafaret
https://github.com/datarobot/datarobot-user-models/blob/3017dca9ef26b4b6b9386a561e7b5062e14ce00f/custom_model_runner/datarobot_drum/runtime_parameters/runtime_parameters_schema.py#L56-L64

and then there is 
https://github.com/datarobot/datarobot-user-models/blob/3b2670a9e2c3f8daca98bafa4b5fcaa498a13d06/custom_model_runner/datarobot_drum/custom_task_interfaces/user_secrets.py#L212-L221

But there is no restriction on which `credential_type` is used by an user.

Reply:
Validation is done in the platform -- there really isn't much benefit in having the validation logic duplicated in DRUM at runtime so we don't both for the most part. See https://github.com/datarobot/DataRobot/blob/master/raptor_lake/custom_model/runtime_parameters.py#L381-L408

Reply:
that is nice, thank you!

Reply:
Done
- removed this check from code
- added the `credentialType` into the `model-metadata.yaml`:
https://github.com/datarobot/datarobot-user-models/blob/882835887839fe98a3b988400577b82254b02a46/model_templates/gpu_nemo_textgen/custom_model/model-metadata.yaml#L10-L13